src/openforms/submissions/tests/test_price_calculation.py

```diff
@@ -0,0 +1,139 @@
+            form__generate_minimal_setup=True,
+            form__product__price=Decimal("123.45"),
+            form__payment_backend="demo",
+            form__price_variable_key="",
+        )
+        FormVariableFactory.create(
+            key="test-key",
+            form=submission.form,
+            form_definition=submission.form.formstep_set.get().form_definition,
+        )
+        SubmissionStepFactory.create(
+            submission=submission,
+            form_step=submission.form.formstep_set.get(),
+            data={"test-key": "test"},
+        )
+        FormPriceLogicFactory.create(
+            form=submission.form,
+            json_logic_trigger={"==": [{"var": "test-key"}, "test"]},
+            price=Decimal("51.15"),
+        )
+        with self.subTest(part="check data setup"):
+            self.assertTrue(submission.payment_required)
+
+        price = get_submission_price(submission)
+
+        self.assertEqual(price, Decimal("51.15"))
+
+    def test_price_from_form_variable(self):
+        submission = SubmissionFactory.create(
+            completed=True,
+            form__generate_minimal_setup=True,
+            form__product__price=Decimal("123.45"),
+            form__payment_backend="demo",
+            form__price_variable_key="",
+        )
+        FormVariableFactory.create(
+            user_defined=True,
+            key="calculatedPrice",
+            form=submission.form,
+            data_type=FormVariableDataTypes.float,
+            # XXX this doesn't play nice with float vs. decimal, but it is what it is
+            initial_value=420.69,
+        )
+        submission.form.price_variable_key = "calculatedPrice"
+        submission.form.save()
+        submission.refresh_from_db()
+        with self.subTest(part="check data setup"):
+            self.assertTrue(submission.payment_required)
+
+        price = get_submission_price(submission)
+
+        self.assertEqual(price, Decimal(420.69))
+
+    def test_price_from_form_variable_error_cases(self):
+        # Variable references can be broken in various ways, in those situation we
+        # expected hard crashes.
+        with self.subTest("broken variable reference"):
+            submission = SubmissionFactory.create(
+                completed=True,
+                form__generate_minimal_setup=False,
+                form__product__price=Decimal("123.45"),
+                form__payment_backend="demo",
+                form__price_variable_key="",
+            )
+            submission.form.price_variable_key = "badVariable.reference"
+            submission.form.save()
+            submission.refresh_from_db()
+
+            with self.assertRaisesMessage(
+                InvalidPrice,
+                "No variable 'badVariable.reference' present in the submission data, "
+                "refusing the temptation to guess.",
+            ):
+                get_submission_price(submission)
+
+        invalid_values = (
```

Comment:
Don't we need the other data types as well?I guess because they would be of the same type (str)?

Reply:
The rationale behind these types was:

* `boolean` is a subclass of `int`, so that needs special treatment on our end
* `string` *could* be in some cases valid, if it's a number like `123.45`, which can be cast to a decimal, but non-numeric strings will cause crashes
* the implementation applies allow-list rather than block-list logic, so the test is geared towards those edge cases. other types could be included, but the details about how the initial value is changed depending on the data type are a bit tricky and peculiar and distract from what's actually being tested, so I opted to leave those out